### PR TITLE
[gptq] inference change replace linear to shardformer

### PR DESCRIPTION
**Pull request by Xu-Kai**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?
- Change gptq tensor parallel replace linear to Shardformer.
- Change colossalai/gptq to colossalai/inference/quant/gptq.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/gptq/cai_gptq/cai_quant_linear.py [hunk 1]
```diff
@@ -2,11 +2,16 @@
 
 import math
 import warnings
+from typing import Callable, List, Optional, Tuple, Union
 
 import numpy as np
```
Comment (yuanheng-zhao) [inline colossalai/gptq/cai_gptq/cai_quant_linear.py:7]: Remove unused imports like lumpy

File: colossalai/gptq/cai_gptq/__init__.py [hunk 2]
```diff
@@ -9,6 +9,6 @@
     HAS_AUTO_GPTQ = False
```
Comment (yuanheng-zhao) [inline colossalai/gptq/cai_gptq/__init__.py:10]: (Just wonder) if we would like to keep the current dir structure (e.g. colossalai.gptq). Refer to Coati inference file structure, `Chat/coati/quant/llama_gptq/`

Comment (Xu-Kai) [reply colossalai/gptq/cai_gptq/__init__.py:10]: how about colossalai/inference/quant/gptq?

File: colossalai/gptq/cai_gptq/cai_quant_linear.py (hunk 1 above)
Comment (Xu-Kai) [reply colossalai/gptq/cai_gptq/cai_quant_linear.py:7]: numpy is used to init.

File: colossalai/inference/tensor_parallel/policies/bloom.py [hunk 3]
```diff
@@ -40,6 +41,33 @@ def module_policy(self):
         if not self.shard_config.inference_gptq:
             policy = super().module_policy()
         else:
+            policy[BloomBlock] = ModulePolicyDescription(attribute_replacement={
```
Comment (yuanheng-zhao) [inline colossalai/inference/tensor_parallel/policies/bloom.py:44]: Do we need to handle `lm_head` when using gptq? `super().module_policy()` here execute `LlamaPolicy.module_policy` and then `LlamaForCausalLMPolicy.module_policy`. https://github.com/hpcaitech/ColossalAI/blob/ce97790ed73f7962ab1ceae057a020168b45dda4/colossalai/shardformer/policies/llama.py#L202-L209C1

File: colossalai/inference/tensor_parallel/policies/bloom.py [hunk 4]
```diff
@@ -40,6 +41,33 @@ def module_policy(self):
...
+                "self_attention.hidden_size": self.model.config.hidden_size // self.shard_config.tensor_parallel_size,
+                "self_attention.split_size": self.model.config.hidden_size // self.shard_config.tensor_parallel_size,
+                "self_attention.num_heads": self.model.config.n_head // self.shard_config.tensor_parallel_size,
+            },
+                                                         sub_module_replacement=[
+                                                             SubModuleReplacementDescription(
+                                                                 suffix="self_attention.query_key_value",
+                                                                 target_module=ColCaiQuantLinear,
```
Comment (CjhHa1) [inline colossalai/inference/tensor_parallel/policies/bloom.py:50]: why the split num are all set fixed

File: colossalai/inference/tensor_parallel/policies/bloom.py (hunk 3 above)
Comment (Xu-Kai) [reply colossalai/inference/tensor_parallel/policies/bloom.py]: now only update decoder block after call uper().module_policy().

File: colossalai/inference/tensor_parallel/policies/bloom.py (hunk 4 above)
Comment (Xu-Kai) [reply colossalai/inference/tensor_parallel/policies/bloom.py:50]: qkv fused linears are 3. others are 1

**Review by CjhHa1 — APPROVED** [review]

**Pull request merged**